OpenRA.Mods.Common/Traits/World/ShroudRenderer.cs

```diff
@@ -234,61 +238,52 @@ Edges GetEdges(PPos puv, Func<PPos, bool> isVisible)
 			return info.UseExtendedIndex ? edge ^ ucorner : edge & Edges.AllCorners;
 		}
 
-		void DirtyCells(IEnumerable<PPos> cells)
+		void WorldOnRenderPlayerChanged(Player player)
 		{
-			// PERF: Many cells in the shroud change every tick. We only track the changes here and defer the real work
-			// we need to do until we render. This allows us to avoid wasted work.
-			cellsDirty.UnionWith(cells);
-		}
+			var newShroud = player != null ? player.Shroud : null;
 
-		void IRenderShroud.RenderShroud(Shroud shroud, WorldRenderer wr)
-		{
-			if (currentShroud != shroud)
+			if (shroud != newShroud)
 			{
-				if (currentShroud != null)
-					currentShroud.CellsChanged -= DirtyCells;
-
 				if (shroud != null)
-					shroud.CellsChanged += DirtyCells;
+					shroud.OnShroudChanged -= UpdateShroudCell;
 
-				// Needs the anonymous function to ensure the correct overload is chosen
-				if (shroud != null)
-					visibleUnderShroud = puv => currentShroud.IsExplored(puv);
+				if (newShroud != null)
+				{
+					visibleUnderShroud = puv => newShroud.IsExplored(puv);
+					visibleUnderFog = puv => newShroud.IsVisible(puv);
+					newShroud.OnShroudChanged += UpdateShroudCell;
+				}
 				else
+				{
 					visibleUnderShroud = puv => map.Contains(puv);
-
-				if (shroud != null)
-					visibleUnderFog = puv => currentShroud.IsVisible(puv);
-				else
 					visibleUnderFog = puv => map.Contains(puv);
+				}
 
-				currentShroud = shroud;
-				DirtyCells(map.ProjectedCellBounds);
+				shroud = newShroud;
 			}
 
-			// We need to update newly dirtied areas of the shroud.
-			// Expand the dirty area to cover the neighboring cells, since shroud is affected by neighboring cells.
-			foreach (var uv in cellsDirty)
-			{
-				cellsAndNeighborsDirty.Add(uv);
-				var cell = ((MPos)uv).ToCPos(map);
-				foreach (var direction in CVec.Directions)
-					cellsAndNeighborsDirty.Add((PPos)(cell + direction).ToMPos(map));
-			}
+			foreach (var uv in map.AllCells.MapCoords)
+				UpdateShroudCell(uv);
+		}
 
-			foreach (var puv in cellsAndNeighborsDirty)
+		void IRenderShroud.RenderShroud(WorldRenderer wr)
+		{
+			foreach (var uv in map.AllCells.MapCoords)
 			{
-				var uv = (MPos)puv;
-				if (!tileInfos.Contains(uv))
+				var uvp1 = new MPos(uv.U + 2, uv.V + 2);
```

Comment:
`uvp1` = `uv` plus 1? Why are we adding 2?

Reply:
This should be called `uvp2` now. This was changed in response to the crash reported by teinarrss. We need to have a wider margin to account for isometric maps.

Reply:
Can you please explain what these are doing? It looks like they are offsetting everything in the `CellLayer` diagonally by two steps, which I can't make sense of.

Reply:
This cellLayer also has four more columns and rows than usual, so the +2 is centering it to obtain a two-cell margin all around.

Reply:
No longer relevant.